### PR TITLE
M5

### PR DESCRIPTION
**Pull request by rgozim**
Various changes:

1. In order to give this plugin the same functionality as the functionality it provides to other gradle plugins, some code is shared between _buildsrc_ and the actual _src/main/kotlin_ folder (using symlinks to avoid IDE errors).

2. Cleaned up some common code into functions in publishingPlugin

3. Added publishing options for binary, or binary and sources.

### CODE REVIEW

File: buildSrc/src/main/kotlin/org/openmicroscopy/PluginPublishingPlugin.kt [hunk 1]
```diff
@@ -0,0 +1,79 @@
+package org.openmicroscopy
+
+import groovy.lang.GroovyObject
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.plugins.GroovyPlugin
+import org.gradle.api.publish.PublishingExtension
+import org.gradle.api.publish.maven.MavenPublication
+import org.gradle.api.publish.maven.plugins.MavenPublishPlugin
+import org.gradle.kotlin.dsl.*
```
Comment (jburel) [inline buildSrc/src/main/kotlin/org/openmicroscopy/PluginPublishingPlugin.kt:10]: only import packages needed

File: buildSrc/src/main/kotlin/org/openmicroscopy/AdditionalArtifactsPlugin.kt [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+import org.gradle.api.Project
+import org.gradle.api.plugins.GroovyPlugin
+import org.gradle.api.plugins.JavaPlugin
+import org.gradle.api.tasks.SourceSet
+import org.gradle.api.tasks.SourceSetContainer
+import org.gradle.api.tasks.bundling.Jar
+import org.gradle.api.tasks.javadoc.Groovydoc
+import org.gradle.api.tasks.javadoc.Javadoc
+import org.gradle.external.javadoc.StandardJavadocDocletOptions
+import org.gradle.kotlin.dsl.*
```
Comment (jburel) [inline buildSrc/src/main/kotlin/org/openmicroscopy/AdditionalArtifactsPlugin.kt:15]: only import packages needed

File: build.gradle.kts [hunk 3]
```diff
@@ -59,62 +52,9 @@ gradlePlugin {
...
-
-fun resolveProperty(envVarKey: String, projectPropKey: String): String? {
-    val propValue = System.getenv(envVarKey)
-    if (propValue != null) {
-        return propValue
-    }
-
-    return findProperty(projectPropKey)?.toString()
-}
+}
```
Comment (jburel) [inline build.gradle.kts]: add extra line

File: buildSrc/build.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+gradlePlugin {
+    plugins {
+        register("plugin-project-plugin") {
+            id = "org.openmicroscopy.plugin-project"
+            implementationClass = "org.openmicroscopy.PluginProjectPlugin"
+        }
+    }
+}
```
Comment (jburel) [inline buildSrc/build.gradle.kts]: same

File: buildSrc/src/main/kotlin/org/openmicroscopy/AdditionalArtifactsPlugin.kt [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+            val groovydoc by tasks.existing(Groovydoc::class)
+
+            tasks.register<Jar>("groovydocJar") {
+                description = "Creates a jar of groovy docs, classified -groovydoc"
+                archiveClassifier.set("groovydoc")
+                from(groovydoc)
+            }
+        }
+    }
+}
```
Comment (jburel) [inline buildSrc/src/main/kotlin/org/openmicroscopy/AdditionalArtifactsPlugin.kt:54]: same

File: buildSrc/src/main/kotlin/org/openmicroscopy/PluginHelper.kt [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+import org.gradle.api.Action
+import org.gradle.api.Project
+import org.gradle.api.artifacts.Configuration
+import org.gradle.api.artifacts.dsl.RepositoryHandler
+import org.gradle.api.artifacts.repositories.ArtifactRepository
+import org.gradle.api.artifacts.repositories.MavenArtifactRepository
+import org.gradle.api.credentials.HttpHeaderCredentials
+import org.gradle.api.plugins.JavaPlugin
+import org.gradle.api.publish.maven.MavenPom
+import org.gradle.kotlin.dsl.*
```
Comment (jburel) [inline buildSrc/src/main/kotlin/org/openmicroscopy/PluginHelper.kt:13]: only import packages needed

File: src/main/kotlin/org/openmicroscopy/tasks/BuildConfig.kt [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+                .addMethod(main)
+                .build()
+
+        val javaFile = JavaFile.builder("com.example.helloworld", helloWorld)
+                .build()
+
+        javaFile.writeTo(System.out)
+    }
+
+}
```
Comment (jburel) [inline src/main/kotlin/org/openmicroscopy/tasks/BuildConfig.kt]: same

File: src/main/kotlin/org/openmicroscopy/tasks/ClassField.kt [hunk 8]
```diff
@@ -0,0 +1,4 @@
+package org.openmicroscopy.tasks
+
+class ClassField {
+}
```
Comment (jburel) [inline src/main/kotlin/org/openmicroscopy/tasks/ClassField.kt]: same

**Comment by joshmoore** [comment]
Two quick comments:
 * In general, I'd prefer to see `WIP` comments rebased away
 * M5 might not be the time, but I again **strongly** suggest that you work towards migrating this repository to groovy (away from kotlin).

**Comment by jburel** [comment]
This works as expected. Follow up PR for copyright etc.

**Pull request merged**